curriculum/challenges/english/25-front-end-development/lab-hash-table/67ed03ac474c48692f41749e.md

```diff
@@ -0,0 +1,400 @@
+
+6. The `lookup` method should:
+
+    - Take a key as its argument.
+    - Compute the hash of the key, and return the corresponding value stored inside the hash table. 
+    - If the key does not exist in the collection, it should return `None`.
+
+# --hints--
+
+You should define a `HashTable` class.
+
+```js
+({ 
+    test: () => assert(runPython(`
+    _Node(_code).has_class("HashTable")
+    `)) 
+})
+```
+
+When a new instance of the `HashTable` class is created, its `collection` attribute should be initialized to an empty dictionary.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        assert ht.collection == {}
+    `) 
+})
+```
+
+The `HashTable` class should have a `hash` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('hash')`))}})
+```
+
+The `hash` method should take a string as a parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.hash)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+The `hash` method should take a string as its argument and return the sum of the Unicode values of each character in the string.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        hash_result = ht.hash("fcc")
+        assert hash_result == 300
+
+        # prevent hardcoding
+        assert ht.hash("golf") == 424
+        assert ht.hash("read") == 412
+`) })
+```
+
+The `HashTable` class should have an `add` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('add')`))}})
+```
+
+The `add` method should take a key and value as parameters.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.add)
+    len(sig.parameters) == 2
+  `))
+})
+```
+
+The `HashTable` class should have a `remove` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('remove')`))}})
+```
+
+The `remove` method should take a key as a parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.remove)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+When you try to remove a key that does not exist in the collection, it should not raise an error or remove anything.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        ht.add("rose", "flower")
+        index = ht.hash("rose")
+        original = ht.collection.copy()
+
+        ht.remove("tulip")
+
+        assert ht.collection == original
+        assert "rose" in ht.collection[index]
+        assert "tulip" not in ht.collection.get(index, {})
+    `) 
+})
+```
+
+
+When the hash value is same, the `remove` method should only remove that key and not the entire hash table for that hash value.
+
+
+```js
+({
+  test: () => runPython(`
+    ht = HashTable()
```

Comment:
The hint says you should only remove the key, but this is actually checking that the hashed value is removed from the table. Also, I would expect two key-value pairs stored at the same hash value in this test.

Not sure if we need to change the hint, the test, or both.

Reply:
I have reworded the hint a bit and added two key-value pairs with the same hash.